lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/group/CachingGroupProviderModule.java

```diff
@@ -0,0 +1,183 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base.group;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import com.google.inject.Binder;
+import com.google.inject.BindingAnnotation;
+import com.google.inject.Key;
+import com.google.inject.Module;
+import com.google.inject.Scopes;
+import io.airlift.configuration.AbstractConfigurationAwareModule;
+import io.trino.spi.security.GroupProvider;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.Optional;
+
+import static io.airlift.configuration.ConditionalModule.conditionalModule;
+import static io.airlift.configuration.ConfigBinder.configBinder;
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * If added to the list of {@link com.google.inject.Module}s used in initialization of a Guice context in a
+ * {@link io.trino.spi.security.GroupProviderFactory}, it will (almost) automatically add caching capability to the
+ * group provider. Requirements:
+ * <ul>
+ *     <li>The {@link GroupProvider} available in the Guice context must be bound annotated with
+ *     {@link ForCachingGroupProvider} binding annotation</li>
+ * </ul>
+ * The module will make the following configuration options available (to be set in {@code etc/group-provider.properties}:
+ * <ul>
+ *     <li>{@code cache.enabled} - the toggle to enable or disable caching</li>
+ *     <li>{@code cache.ttl} - determines how long group information will be cached for each user</li>
+ *     <li>{@code cache.maximum-size} - maximum number of users for which groups are stored in the cache</li>
+ * </ul>
+ * These properties can optionally have an arbitrary prefix ({@link Builder#withPrefix(String)})
+ * and/or a binding annotation for the resulting binding of {@link GroupProvider} ({@link Builder#withBindingAnnotation(Class)}).
+ * <p>
+ * An additional object of type {@link GroupCacheInvalidationController} will also be bound, with which one can invalidate
+ * all or part of the cache.
+ */
+public class CachingGroupProviderModule
+        extends AbstractConfigurationAwareModule
+{
+    private final Optional<String> prefix;
+    private final Optional<Class<? extends Annotation>> bindingAnnotation;
+
+    private CachingGroupProviderModule(Optional<String> prefix, Optional<Class<? extends Annotation>> bindingAnnotation)
```

Comment:
Do we need the `Prefix` and annotations based approach here ? Each `GroupProviderFactory` would provide only a single `GroupProvider` so bootstrap doesn't have to worry about multiple GroupProvider being injected - or am I missing something here ?

cc: @ksobolew

Reply:
I guess the idea was that this would be a generic caching layer that could be applied to any group provider (it has nothing that is specific to LDAP). In that case allowing a configurable prefix makes sense. But perhaps we can consider this a case of premature optimization for not-yet-existing use cases.

Reply:
And yes, there would be only one group provider per factory, but I thought that it looks nicer when you have config like:

```
ldap.something = aaaa
ldap.something-else = bbb
ldap.cache.something = ccc
```

instead of:

```
ldap.something = aaaa
ldap.something-else = bbb
cache.something = ccc
```

Reply:
In this case having a prefix would help right ? I''m a bit confused if we need Annotations specific binding.

Reply:
Ah, yes, the qualifying annotation. I don't remember why I though it was needed, sorry :)